From f5fa81dba36f2e45ff96ec88187557f04c02ceb1 Mon Sep 17 00:00:00 2001
From: Lars Schmertmann <Lars.Schmertmann@governikus.de>
Date: Fri, 22 Jan 2021 07:41:28 +0100
Subject: Allow QNearFieldTarget to detect a lost card

Some devices (Samsung, Huawei) throw an exception when the card is lost:
"android.nfc.TagLostException: Tag was lost". But there seems to be a bug that
isConnected still reports true. So we need to invalidate the target to allow
checkIsTargetLost to recognize this missing card.

Pick-to: 5.15
Change-Id: Ied95e8b906ea19e2e289f28af736ab52e2221234
Reviewed-by: Alex Blasche <alexander.blasche@qt.io>
(cherry picked from commit f5ea7dc7248833af95b092a94e7cc9e666705200)
---
 src/nfc/qnearfieldtarget_android.cpp | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git x/qtconnectivity/src/nfc/qnearfieldtarget_android.cpp y/qtconnectivity/src/nfc/qnearfieldtarget_android.cpp
index 16aafbf1..78468bf6 100644
--- x/qtconnectivity/src/nfc/qnearfieldtarget_android.cpp
+++ y/qtconnectivity/src/nfc/qnearfieldtarget_android.cpp
@@ -253,6 +253,12 @@ QNearFieldTarget::RequestId NearFieldTarget::sendCommand(const QByteArray &comma
     // Writing
     QAndroidJniObject myNewVal = m_tagTech.callObjectMethod("transceive", "([B)[B", jba);
     if (catchJavaExceptions()) {
+        // Some devices (Samsung, Huawei) throw an exception when the card is lost:
+        // "android.nfc.TagLostException: Tag was lost". But there seems to be a bug that
+        // isConnected still reports true. So we need to invalidate the target to allow
+        // checkIsTargetLost to recognize this missing card.
+        m_intent = QAndroidJniObject();
+
         reportError(QNearFieldTarget::CommandError, requestId);
         return requestId;
     }
-- 
2.33.0

